### DISCUSSION TITLE
OPA Wasm Compilation - Automatic Entrypoint Detection

### DISCUSSION THREAD

**Discussion by khaledbaghdadii** [open-policy-agent/community#448] [OPA and Rego]
I'm working on a project that involves compiling a bunch of Rego policies into Wasm bundles. The thing is, I've got loads of policy files in nested directories and specifying the entrypoint for each one is a real headache.

Any chance there's a feature or trick that lets OPA auto-detect the entrypoints?

**Comment by anderseknert** [comment]
Hi there! I'm not sure how it would be able to do that 😅 Any rule or package is potentially an entrypoint, and while OPA _maybe_ could guess the "main" entrypoint by... I don't know, count the number of dependencies to other rules or something like that, you're specifically saying that you have a _bunch_ of them. 

Could you expand on why you need so many entrypoints? Can't you just have a single one that routes to the rules needed for evaluation? [Dynamic policy composition](https://www.styra.com/blog/dynamic-policy-composition-for-opa/) is a nice pattern to work with for that purpose.

**Reply by khaledbaghdadii** [reply]
Hello, thanks for the reply.
I'll check the article provided in the answer. As of why we need so many endpoints, in our project we have multiple resources and multiple actions that can be executed on each one, and we have a rego file defined for each resource.

**Reply by anderseknert** [reply]
Having a main policy / entrypoint that "routes" to other policies and rules based on the content in input seems like a good idea then 🙂

**Comment by srenatus** [comment]
Would using `entrypoint: true` in metadata not help here? You'd only have to annotate all the rules where they are written, you don't have to maintain a second list for the `opa built -t wasm` call.

**Reply by khaledbaghdadii** [reply]
That might be useful actually, I'll check it out. Thanks!
